flang/test/Lower/forall-2.f90

```diff
@@ -15,20 +15,53 @@ subroutine implied_iters_allocatable(a1)
 end subroutine implied_iters_allocatable
 
 ! CHECK-LABEL: func @_QPforall_pointer_assign(
+! CHECK-SAME: %[[arg0:[^:]*]]: !fir.box<!fir.array<?x!fir.type<_QFforall_pointer_assignTt{ptr:!fir.box<!fir.ptr<!fir.array<?xf32>>>}>>>,
+! CHECK-SAME: %[[arg1:[^:]*]]: !fir.box<!fir.array<?x!fir.type<_QFforall_pointer_assignTu{targ:!fir.array<20xf32>}>>> {fir.target},
+! CHECK-SAME: %[[arg2:[^:]*]]: !fir.ref<i32>, %[[arg3:[^:]*]]: !fir.ref<i32>) {
 subroutine forall_pointer_assign(ap, at, ii, ij)
+  ! CHECK-NEXT: %[[VAL_0:.*]] = fir.alloca i32 {adapt.valuebyref, uniq_name = "i"}
+  ! CHECK: %[[VAL_1:.*]] = fir.array_load %[[arg0]] : (!fir.box<!fir.array<?x!fir.type<_QFforall_pointer_assignTt{ptr:!fir.box<!fir.ptr<!fir.array<?xf32>>>}>>>) -> !fir.array<?x!fir.type<_QFforall_pointer_assignTt{ptr:!fir.box<!fir.ptr<!fir.array<?xf32>>>}>>
+  ! CHECK: %[[VAL_3:.*]] = fir.array_load %[[arg1]] : (!fir.box<!fir.array<?x!fir.type<_QFforall_pointer_assignTu{targ:!fir.array<20xf32>}>>>) -> !fir.array<?x!fir.type<_QFforall_pointer_assignTu{targ:!fir.array<20xf32>}>>
+  ! CHECK: %[[VAL_5:.*]] = fir.load %[[arg2]] : !fir.ref<i32>
+  ! CHECK: %[[VAL_7:.*]] = fir.convert %[[VAL_5]] : (i32) -> index
+  ! CHECK: %[[VAL_8:.*]] = fir.load %[[arg3]] : !fir.ref<i32>
+  ! CHECK: %[[VAL_10:.*]] = fir.convert %[[VAL_8]] : (i32) -> index
+  ! CHECK: %[[VAL_11:.*]] = constant 8 : i32
+  ! CHECK: %[[VAL_12:.*]] = fir.convert %[[VAL_11]] : (i32) -> index
+
   type t
      real, pointer :: ptr(:)
   end type t
   type u
      real :: targ(20)
   end type u
   type(t) :: ap(:)
-  type(u) :: at(:)
+  type(u), target :: at(:)
   integer :: ii, ij
 
+  ! CHECK: %[[V_0:.*]] = fir.do_loop %[[V_1:.*]] = %[[VAL_7]] to %[[VAL_10]] step %[[VAL_12]] unordered iter_args(%[[VAL_2:.*]] = %[[VAL_1]]) -> (!fir.array<?x!fir.type<_QFforall_pointer_assignTt{ptr:!fir.box<!fir.ptr<!fir.array<?xf32>>>}>>) {
+  ! CHECK: %[[V_3:.*]] = fir.convert %[[V_1]] : (index) -> i32
+  ! CHECK: fir.store %[[V_3]] to %[[VAL_0]] : !fir.ref<i32>
+  ! CHECK: %[[VAL_4:.*]] = fir.field_index targ, !fir.type<_QFforall_pointer_assignTu{targ:!fir.array<20xf32>}>
+  ! CHECK-DAG: %[[VAL_5:.*]] = fir.load %[[VAL_0]] : !fir.ref<i32>
+  ! CHECK-DAG: %[[VAL_6:.*]] = constant 4 : i32
+  ! CHECK: %[[VAL_7:.*]] = subi %[[VAL_5]], %[[VAL_6]] : i32
+  ! CHECK: %[[VAL_8:.*]] = fir.convert %[[VAL_7]] : (i32) -> i64
+  ! CHECK: %[[VAL_9:.*]] = fir.convert %[[VAL_8]] : (i64) -> index
+  ! CHECK: %[[VAL_10:.*]] = fir.array_fetch %[[VAL_3]], %[[VAL_9]], %[[VAL_4]] : (!fir.array<?x!fir.type<_QFforall_pointer_assignTu{targ:!fir.array<20xf32>}>>, index, !fir.field) -> !fir.ref<!fir.array<20xf32>>
+  ! CHECK: %[[VAL_11:.*]] = fir.embox %[[VAL_10]] : (!fir.ref<!fir.array<20xf32>>) -> !fir.box<!fir.array<20xf32>>
```

Comment:
Good point on the type. I left a TODO/FIXME somewhere as, yes, I still need to refactor things since the way scalar assignments are lowered isn't really amenable to array contexts yet.